### ISSUE TITLE
Potential bug in ag_expand

### ISSUE THREAD

**Issue by meng-li** [dariasor/TreeExtra#3]
if i run *ag_train* command with following parameters,
``` shell
ag_train -t data.train -v data.valid -r data.attr -a 0.01
```
it will suggest to get better result with 
```shell
ag_expand -a 0 -b 140
```
this may caused a potential bug in *ag_expand* part. Since in the previous iteration, alpha walk through **0.5, 0.2, 0.1, 0.05, 0.02, 0.01**, in current iteration,*alpha* will walk through **0.5, 0.2, 0.1, 0.05, 0.02, 0.01, 0.0**. This will bring a potential bug since **AGTemp/ag.a.0.0.n.\*.tmp* file haven't been created,  [following code](https://github.com/dariasor/TreeExtra/blob/master/AdditiveGroves/ag_expand.cpp#L338) will throw a **TEMP_ERR**. 
```c++ [https://github.com/dariasor/TreeExtra/blob/master/AdditiveGroves/ag_expand.cpp#L338](url)
if(bagNo == 0)
    pftemp = new fstream(tempFName.c_str(), ios_base::binary | ios_base::in);
if(pftemp->fail())
    throw TEMP_ERR;	
```
Maybe i misunderstand some logic, please help me to fix it, thanks!

**Comment by dariasor** [comment]
It is not supposed to crash, because this part of code is under the following if clause:
					if((tigNNo == prevTiGNN - 1) && (alphaNo < prevAlphaN) ||
					   (alphaNo == prevAlphaN - 1) && (tigNNo < prevTiGNN))

So if the previous iteration of ag_train/ag_expand did not go all the way up to alpha=0, it won't execute.

But if you are able to actually crash ag_expand, please send me the log.txt and I'll definitely debug.

**Comment by meng-li** [comment]
Thank you for helping to fix this problem.

> Environment
> + g++ (Ubuntu 4.3.3-5ubuntu4) 4.3.3
> + system: Linux localhost.localdomain 2.6.28-11-server #42-Ubuntu SMP Fri Apr 17 02:48:10 UTC 2009 i686 GNU/Linux

[log.txt](https://github.com/dariasor/TreeExtra/files/972331/log.txt)


-----
ag_train -t /home/limeng/TreeExtra-master/data/data.train -v /home/limeng/TreeExtra-master/data/data.valid -r /home/limeng/TreeExtra-master/data/data.attr -s fast 

Reading the attribute file: "/home/limeng/TreeExtra-master/data/data.attr"
10 attributes
10 active attributes

Reading the train set: "/home/limeng/TreeExtra-master/data/data.train"
100 points in the train set, std. dev. of y values = 0.911907

Reading the validation set: "/home/limeng/TreeExtra-master/data/data.valid"
100 points in the validation set, std. dev. of y values = 0.89671

Alpha = 0.01
N = 8
60 bagging iterations
fast mode

Best model:
	Alpha = 0.1
	N = 8
RMSE on validation set = 0.264684

Recommendation: relaxing model parameters might produce a better model.
 Suggested action: ag_expand -n 16 -b 100


-----
ag_expand -n 16 -b 100 

Reading the attribute file: "/home/limeng/TreeExtra-master/data/data.attr"
10 attributes
10 active attributes

Reading the train set: "/home/limeng/TreeExtra-master/data/data.train"
100 points in the train set, std. dev. of y values = 0.911907

Reading the validation set: "/home/limeng/TreeExtra-master/data/data.valid"
100 points in the validation set, std. dev. of y values = 0.89671

Alpha = 0.01
N = 16
100 bagging iterations
fast mode

Previous model's RMSE on validation set = 0.264684

Best model:
	Alpha = 0.1
	N = 11
RMSE on validation set = 0.25011

Recommendation: relaxing model parameters might produce a better model.
 Suggested action: ag_expand -b 140


-----
ag_expand -b 140 

Reading the attribute file: "/home/limeng/TreeExtra-master/data/data.attr"
10 attributes
10 active attributes

Reading the train set: "/home/limeng/TreeExtra-master/data/data.train"
100 points in the train set, std. dev. of y values = 0.911907

Reading the validation set: "/home/limeng/TreeExtra-master/data/data.valid"
100 points in the validation set, std. dev. of y values = 0.89671

Alpha = 0.01
N = 16
140 bagging iterations
fast mode

Previous model's RMSE on validation set = 0.25011

Best model:
	Alpha = 0.01
	N = 16
RMSE on validation set = 0.246136

Recommendation: relaxing model parameters might produce a better model.
 Suggested action: ag_expand -a 0 -n 32 -b 180


-----
ag_expand -a 0 -n 32 -b 180 

Reading the attribute file: "/home/limeng/TreeExtra-master/data/data.attr"
10 attributes
10 active attributes

Reading the train set: "/home/limeng/TreeExtra-master/data/data.train"
100 points in the train set, std. dev. of y values = 0.911907

Reading the validation set: "/home/limeng/TreeExtra-master/data/data.valid"
100 points in the validation set, std. dev. of y values = 0.89671

Alpha = 0
N = 32
180 bagging iterations
fast mode

Previous model's RMSE on validation set = 0.246136

Error: temporary files from previous runs of train/expand are missing or corrupted.

**Comment by dariasor** [comment]
Hmm, can't reproduce. When I am running the current development version, I get the following log for ag_train. Note that min alpha is changed to zero, that somehow has not happened in your log.
What version of the TreeExtra code are you using?

ag_train -t data.train -v data.valid -r data.attr -s fast

Reading the attribute file: "data.attr"
10 attributes
10 active attributes

Reading the train set: "data.train"
100 points in the train set, std. dev. of y values = 0.911907

Reading the validation set: "data.valid"
100 points in the validation set, std. dev. of y values = 0.89671

Warning: due to small train set size (100) value of alpha was changed to 0.

Alpha = 0
N = 8
60 bagging iterations
fast mode

Best model:
        Alpha = 0.1
        N = 8
RMSE on validation set = 0.257154

Recommendation: relaxing model parameters might produce a better model.
Suggested action: ag_expand -n 16 -b 100

**Comment by meng-li** [comment]
i just clone this project from **master** branch. But there is something weird happens in my environment, when i use **g++ -O3** to build the project, i will encounter an exception in following part
``` c++
int bestSplitN = (int)bestSplits.size(); (Note: =>bestSplitN will be zero!)
int randSplit = rand() % bestSplitN; 
```
However, if i use **g++ -O0**, it will be ok. 
I made some change in **AdditiveGroves/Makefile**

```shell
SHAREDDIR=../shared
 LIBDIR=../ThreadPool
 CXXFLAGS = -O0 -I$(SHAREDDIR) -I$(LIBDIR)
 OBJS = Grove.o $(SHAREDDIR)/SplitInfo.o  $(SHAREDDIR)/INDdata.o $(SHAREDDIR)/TreeNode.o ag_functions.o $(SHAREDDIR)/functions.o $(SHAREDDIR)/LogStream.o $(LIBDIR)/thread_pool.o
 PGMS = ag_predict ag_train ag_save ag_addbag ag_expand ag_merge ag_fs ag_interactions ag_nway ag_mergepreds ag_savemerge
 PGMOBJS = ag_predict.o ag_train.o ag_save.o ag_addbag.o ag_expand.o ag_merge.o ag_fs.o ag_interactions.o ag_nway.o ag_mergepreds.o ag_savemerge.o
 LIBS = -lpthread


 all: directories $(PGMS)

 directories:
     mkdir -p ../Bin

 clean:
     rm -rf $(OBJS)
     rm -rf $(PGMOBJS)
     rm -rf $(PGMS)
     rm -rf ../Bin/*

 .cpp.o:
     g++ $(CXXFLAGS) -c $< -g -o $@

 ag_predict: ag_predict.o $(OBJS)
     g++ -g $(CXXFLAGS) -o ../Bin/ag_predict ag_predict.o $(OBJS) $(LIBS)

 ag_train: ag_train.o $(OBJS)
     g++ -g $(CXXFLAGS) -o ../Bin/ag_train ag_train.o $(OBJS) $(LIBS)

 ag_save: ag_save.o $(OBJS)
     g++ -g $(CXXFLAGS) -o ../Bin/ag_save ag_save.o $(OBJS) $(LIBS)

 ag_addbag: ag_addbag.o $(OBJS)
     g++ -g $(CXXFLAGS) -o ../Bin/ag_addbag ag_addbag.o $(OBJS) $(LIBS)

 ag_expand: ag_expand.o $(OBJS)
     g++ -g $(CXXFLAGS) -o ../Bin/ag_expand ag_expand.o $(OBJS) $(LIBS)
ag_merge: ag_merge.o $(OBJS)
     g++ -g $(CXXFLAGS) -o ../Bin/ag_merge ag_merge.o $(OBJS)  $(LIBS)

 ag_fs: ag_fs.o $(OBJS)
     g++ -g $(CXXFLAGS) -o ../Bin/ag_fs ag_fs.o $(OBJS) $(LIBS)

 ag_interactions: ag_interactions.o $(OBJS)
     g++ -g $(CXXFLAGS) -o ../Bin/ag_interactions ag_interactions.o $(OBJS) $(LIBS)

 ag_nway: ag_nway.o $(OBJS)
     g++ -g $(CXXFLAGS) -o ../Bin/ag_nway ag_nway.o $(OBJS) $(LIBS)

 ag_mergepreds: ag_mergepreds.o $(OBJS)
     g++ -g $(CXXFLAGS) -o ../Bin/ag_mergepreds ag_mergepreds.o $(OBJS) $(LIBS)

 ag_savemerge: ag_savemerge.o $(OBJS)
     g++ -g $(CXXFLAGS) -o ../Bin/ag_savemerge ag_savemerge.o $(OBJS)    $(LIBS)
```

This problem is really weird, Thank you for helping fix this~

**Comment by dariasor** [comment]
I see, looks like something is breaking during the compilation. I'll try to find a way to test on Ubuntu (I've been working on RHEL for the past few years).

**Comment by meng-li** [comment]
@dariasor OK, thank you in advance~

**Comment by meng-li** [comment]
hi @dariasor , i have read the paper "Additive Groves of Regression Trees", the idea is great! But when i have deep look in this code, your way to deal with "Missing value" confuse me. Can you recommend some resource for me to have a deep understand about this part? Thanks!

**Comment by dariasor** [comment]
Turned out, the old g++ had two issues that showed up for this code: incorrect optimization of NaN comparisons and issues with exact doubles comparisons. Both problems are fixed - the code does not have the problematic direct comparisons any more - and the changes are merged into master branch as of 05/30/2017. Meng confirmed that the improved code works for his environment as expected. Closing the issue.
